feat(options)!: deprecate paste, remove pastetoggle

ii14 opened the pull request.

justinmk commented:
we cannot remove 'paste'. It is very common in plugins and configs.

'pastetoggle' can and should be removed though, it's a total waste of everyone's time because it generates bug reports and doesn't work well, and is useless because bracketed-paste works better.

ii14 committed:
feat(options)!: deprecate paste, remove pastetoggle

ii14 committed:
fixup! docs: remove references to paste

justinmk commented on runtime/doc/deprecated.txt:
```diff
@@ -145,6 +145,7 @@ OPTIONS
 - 'viewoptions'		Flags "unix", "slash" are ignored and always enabled.
 - *'viminfo'*		Deprecated alias to 'shada' option.
 - *'viminfofile'*	Deprecated alias to 'shadafile' option.
+- *'paste'* *'nopaste'*
```
```
	This option is obsolete; |paste| is handled automatically.
	Enables "paste mode":
		- mappings in Insert mode and Command-line mode are disabled
		- abbreviations are disabled
		- 'autoindent' is reset
		- 'expandtab' is reset
		- 'formatoptions' is used like it is empty
		- 'revins' is reset
		- 'ruler' is reset
		- 'showmatch' is reset
		- 'smartindent' is reset
		- 'smarttab' is reset
		- 'softtabstop' is set to 0
		- 'textwidth' is set to 0
		- 'wrapmargin' is set to 0
	These options keep their value, but their effect is disabled:
		- 'cindent'
		- 'indentexpr'
		- 'lisp'
```

ii14 replied:
Shouldn't this be under pastetoggle?

justinmk replied:
it applies to both.

Just Paste It.™

ii14 commented on runtime/optwin.vim:
```diff
@@ -230,14 +218,6 @@ call <SID>AddOption("compatible", gettext("behave very Vi compatible (not advisa
 call <SID>BinOptionG("cp", &cp)
 call <SID>AddOption("cpoptions", gettext("list of flags to specify Vi compatibility"))
 call <SID>OptionG("cpo", &cpo)
-call <SID>AddOption("paste", gettext("paste mode, insert typed text literally"))
-call <SID>BinOptionG("paste", &paste)
```
Do we keep deprecated options in optwin?

justinmk replied:
i'd lean towards leaving it in, no strong opinion

justinmk commented on runtime/doc/news.txt:
```diff
@@ -55,6 +55,8 @@ The following changes may require adaptations in user config or plugins.
   - The `concat` option has been removed as it was not consistently applied.
   - Invalid ranges now cause an error instead of returning `nil`.
 
+• 'paste' option is now deprecated and 'pastetoggle' is removed
```
```suggestion
• 'paste' option is now deprecated and 'pastetoggle' is removed. |paste| works
  automatically in GUI and terminal (TUI) Nvim.  Just Paste It.™
```

justinmk commented on runtime/doc/starting.txt:
```diff
@@ -732,8 +732,8 @@ that when these commands are executed, the current key mappings and options
 these are terminal or file dependent.
-Note that the options 'binary', 'paste' and 'readonly' are included, this
-might not always be what you want.
+Note that the options 'binary' and 'readonly' are included, this might not
```
still should mention 'paste' here for now

ii14 committed:
fixup! docs

ii14 committed:
fixup! oldtests

ii14 committed:
fixup! bring back paste to optwin

ii14 committed:
fixup! docs: deprecated

zeertzjq committed:
fixup! oldtests

The test in test_paste.vim can be removed as it is considered N/A by
vim-patch.sh

zeertzjq commented:
'paste' may have to be used internally in the future. #17723

justinmk commented:
> 'paste' may have to be used internally in the future. #17723

yeah, though in theory `nvim_buf_set_text()` or similar should handle this kind of thing. Related: https://github.com/neovim/neovim/issues/19806

zeertzjq committed:
fixup! remove code used only for 'pastetoggle'

zeertzjq committed:
fixup! group removed options together

zeertzjq committed:
fixup! remove more code used only for 'pastetoggle'

zeertzjq committed:
fixup! avoid high CPU usage

zeertzjq committed:
Revert "fixup! avoid high CPU usage"

This reverts commit bddd65d78273c5e023f293be8e5a3313aaeaae8a.

zeertzjq committed:
Revert "fixup! remove more code used only for 'pastetoggle'"

This reverts commit 5209bcae2dceb56f6a3e8d1f9101a84519dcb8f1.

hinell commented:
I don't really like the way it's named. It would be better to have something like `vim.clipboard.get() ... set()`

fredizzimo commented:
Can `paste/nopaste` really be deprecated, or am I missing something here? 

For example, using in insert mode doing `<C-r">`  (or any other register), still pastes with the wrong formatting, Furthermore, not all UI's implement paste, for example Neovide currently does not,  although it probably should, I guess by calling `nvim_paste`?

zeertzjq commented:
CTRL-R CTRL-O

fredizzimo commented:
Thanks, yes that works, I had forgot that mapping since I rarely paste from insert and almost never any longer text. That along with its siblings `<C-R><C-P>` and `<C-R><C-R>` are superior alternatives in most cases.

But I feel like they should be documented as alternatives in the deprecated section though. `set paste` is a common recommendation when you search for "paste in vim insert mode without formatting" in Google. So, it's likely that many people are relying on either that or `pastetoggle`

justinmk commented:
> For example, using in insert mode doing `<C-r">` (or any other register), still pastes with the wrong formatting, 

We should change that. `<c-r>{clipboard}` should work just like `vim.paste()`, it shouldn't try to be "smart".

> Furthermore, not all UI's implement paste, for example Neovide currently does not, although it probably should, I guess by calling `nvim_paste`?

yes, see `:help dev-ui`

9mm commented:
@justinmk is there something special I need to do with `CTRL-R CTRL-P {register}`?

For me, it still isn't fixing the indenting. I have to try to use visual selection and `=` and then its actually fixed.

justinmk commented:
> pasting something "just works", regardless of whether its:
> 
> 1. insert mode, normal mode, visual mode, visual line mode... visual block mode

@9mm that is also the case in Nvim if you use "OS paste". I assume you were using "OS paste" (cmd+v) in macvim. Compare apples to apples.

Doing that for `+"p` and friends is a separate topic. 

This isn't a support thread, use https://github.com/neovim/neovim/discussions for questions.
